VSD - max array length

jezhiggins opened the pull request:
To prevent large arrays getting unwieldy, add a configuration parameter that limits the number of entries they have. 

Writes beyond the maximum index are all written to the same location, merging with any previous writes. 

* Adds `--vsd-arrays smash` option, which sets the maximum size to 1. All writes to the array are merged, and all reads return the same value.
* Add `--vsd-arrays up-to-n-elements` option. Default n is 10, `--vsd-max-array-elements` options gives finer control. (I would have liked n to be an additional parameter to --vsd-arrays, but that would have required significant rework in the arguments parsing and it's just not worth it.)

jezhiggins committed:
Add EXPECT_INDEX & EXPECT_EMPTY_INDEX for more concise array unit tests

jezhiggins committed:
Initial unit tests to capture array write behaviour

jezhiggins committed:
Rework array eval_index to remove out parameter

jezhiggins committed:
array insert/replace helper functions

jezhiggins committed:
Hardwire maximum array size

jezhiggins committed:
merge writes beyond array maximum length

jezhiggins committed:
Remove redundant vsd_configt::advanced_sensitivities

jezhiggins committed:
Export configuration on object factory, and on abstract_environmentt

jezhiggins committed:
Add vsd_configt::maximum_array_index, pick up in full_array_abstract_object

martin-cs commented on src/analyses/variable-sensitivity/variable_sensitivity_configuration.h:
```diff
@@ -51,18 +52,15 @@ struct vsd_configt
 
   flow_sensitivityt flow_sensitivity;
 
+  size_t maximum_array_index = std::numeric_limits<size_t>::max();
```
I presume this is pending future changes to option handling?

jezhiggins replied:
Yes, that's right.

jezhiggins committed:
Combined a similiar couple of tests into one with multiple descs

jezhiggins committed:
Add --vsd-arrays smash option

The "smash" array abstraction sets the maximum size of the
full_array_abstract_object to 1. All writes to the array are merged, and
all reads return the same value.

jezhiggins committed:
Prevent simultaneous data-dependencies and liveness options

jezhiggins committed:
Bigger array in variable-sensitivity-array-nondet-access tests

jezhiggins committed:
Add --vsd-arrays up-to-n-elements option

Default n is 10.
Additional --vsd-max-array-elements options gives finer control.

jezhiggins commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -337,10 +337,9 @@ abstract_object_pointert abstract_environmentt::abstract_object_factory(
 {
-  return object_factory->wrap_with_context(abstract_object);
+  return object_factory->config();
 }
```
Bit of a diff mismatch. `add_object_context` is redundant and was removed. `configuration` is a new member function, which we need :)
